### PR TITLE
Resolves #310: KPI to track occurrences of index entries that do not point to valid records

### LINKED ISSUE #310: KPI to track occurrences of index entries that do not point to valid records
We should have a KPI that tracks each time we hit an index entry for which there is no corresponding record.

### PR DESCRIPTION
**Pull request by nblintao**

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreBase.java [hunk 1]
```diff
@@ -821,6 +821,9 @@ default boolean recordExists(@Nonnull final Tuple primaryKey, @Nonnull final Iso
         final Tuple primaryKey = indexEntryPrimaryKey(index, entry.getKey());
         return loadRecordInternal(primaryKey, false).thenApply(record -> {
             if (record == null) {
+                if (getTimer() != null) {
+                    getTimer().increment(FDBStoreTimer.Counts.BAD_INDEX_ENTRY);
```
Comment (scgray) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreBase.java]: This `if()` block should be moved under the `case ERROR` below.  The `SKIP` and `RETURN` options are used by index maintainers that wish to defer index maintenance until later, so those missing index entries are expected.

Comment (nblintao) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreBase.java]: Done.

**Comment by nblintao** [comment]
```
com.apple.foundationdb.record.provider.foundationdb.keyspace.ExtendedDirectoryLayerTest > testSetMappingWithConflicts() FAILED
    org.opentest4j.AssertionFailedError: nothing was added for the wrong value ==> Expected java.util.concurrent.CompletionException to be thrown, but nothing was thrown
```
test this please

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBStoreTimer.java [hunk 2]
```diff
@@ -420,7 +420,10 @@ public String title() {
         /** The number of times that a leaderboard update deleted a time window. */
         TIME_WINDOW_LEADERBOARD_DELETE_WINDOW("number of leaderboard windows deleted", false),
         /** The number of times that a leaderboard needs to be rebuilt because a window was added after a score it should contain. */
-        TIME_WINDOW_LEADERBOARD_OVERLAPPING_CHANGED("number of leaderboard conditional rebuilds", false);
+        TIME_WINDOW_LEADERBOARD_OVERLAPPING_CHANGED("number of leaderboard conditional rebuilds", false),
+        /** The number of times that a index entry does not point to a valid record. */
```
[the remark this answers is not in the document]
Comment (nblintao) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBStoreTimer.java]: Good catch. Done.

**Comment by alecgrieser** [comment]
@scgray, this looks good to me, but I wasn't actually asked for a review (I'm just ultracrepidarian). Should I merge?

**Comment by scgray** [comment]
Thanks @alecgrieser - I had been waiting for the tests to finish running and then forgot to get back to it.

**Pull request merged**